main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala

```diff
@@ -652,4 +651,235 @@ object SetFormula {
   def mkDifference(f1: SetFormula, f2: SetFormula): SetFormula =
     mkInter(f1, mkCompl(f2))
 
+  //
+  // Other Functions
+  //
+
+  /**
+    * Propagation uses the following rewrites:
+    *   - `x ∩ f == x ∩ f[x -> Univ]`
+    *   - `!x ∩ f == !x ∩ f[x -> Empty]`
+    *   - `x ∪ f == x ∪ f[x -> Empty]`
+    *   - `!x ∪ f == !x ∪ f[x -> Univ]`
+    *
+    * where `x` is [[Cst]], [[Var]], or [ElemSet]].
+    *
+    * Invariant: [[ElemSet]], [[Cst]], and [[Var]] must use disjoint integers.
+    */
+  def propagation(t: SetFormula): SetFormula = propagationHelper(t, SortedMap.empty)
+
+  /**
+    * Applies set propagation to `f` where `insts` keep track of running instantiations for
+    * [[ElemSet]], [[Cst]], and [[Var]] (using the identity mapping if not present).
+    */
+  private def propagationHelper(f: SetFormula, insts: SortedMap[Int, UnivOrEmpty]): SetFormula = f match {
```

Comment:
Perhaps the two functions could be called: propagate and propagateWithEnv/propagateWithInsts?

I always find "helper" suffix to be useless information.

Reply:
It is useless, I just use it to avoid shadowing